File: updated_code/code/python/144-binary-tree-preorder-traversal.py
from typing import Optional, List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    """
    Given the `root` of a binary tree, return the preorder traversal of its nodes' values.
    """
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        """
        This function performs a preorder traversal of a binary tree
        Args:
        root (Optional[TreeNode]): The root of the binary tree

        Returns:
        List[int]: The preorder traversal of the binary tree
        """
        if not root:
            return []

        result = []
        stack = [root]

        while stack:
            node = stack.pop()
            result.append(node.val)

            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)

        return result

File: updated_code/code/java/144-binary-tree-preorder-traversal.java
package com.lawrence;

import java.util.ArrayList;
import java.util.Stack;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    
    /**
     * Returns the preorder traversal of the given binary tree's nodes' values.
     * 
     * @param root the root of the binary tree.
     * @return a list of node values in preorder traversal.
     */
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        
        // Initialize a stack with the root node
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        
        // Perform preorder traversal
        while (!stack.isEmpty()) {
            TreeNode node = stack.pop();
            result.add(node.val);
            
            // Add right and left child to the stack (if exists)
            if (node.right != null) {
                stack.push(node.right);
            }
            if (node.left != null) {
                stack.push(node.left);
            }
        }
        
        return result;
    }
}